<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 4 Sprint 2*

# Sprint Challenge - Neural Network Foundations

Table of Problems

1. [Defining Neural Networks](#Q1)
2. [Chocolate Gummy Bears](#Q2)
    - Perceptron
    - Multilayer Perceptron
4. [Keras MMP](#Q3)

<a id="Q1"></a>
## 1. Define the following terms:

- **Neuron:** A neuron is anytype of the three types of layers: input, hidden or output. Signals flow from early neurons and if the signal passes the threshold, the next neuron will fire off a signal to the next.
- **Input Layer:** Visible layer is the info we get from our dataset and its usually one input node per feature
- **Hidden Layer:** This is the middle layer of nodes that connects to every other node. Typically depth (more layers) is more important than width (more nodes) for neural networks.
- **Output Layer:** The produced result for the given inputs of the neural network
- **Activation:** Or transfer function: The values our equation generates from the previous values it was given-if activated the equation with output the correstponding values.
- **Backpropagation:** Weights in the neural network are updated in reverse order at the end of each training epoch


## 2. Chocolate Gummy Bears <a id="Q2"></a>

Right now, you're probably thinking, "yuck, who the hell would eat that?". Great question. Your candy company wants to know too. And you thought I was kidding about the [Chocolate Gummy Bears](https://nuts.com/chocolatessweets/gummies/gummy-bears/milk-gummy-bears.html?utm_source=google&utm_medium=cpc&adpos=1o1&gclid=Cj0KCQjwrfvsBRD7ARIsAKuDvMOZrysDku3jGuWaDqf9TrV3x5JLXt1eqnVhN0KM6fMcbA1nod3h8AwaAvWwEALw_wcB). 

Let's assume that a candy company has gone out and collected information on the types of Halloween candy kids ate. Our candy company wants to predict the eating behavior of witches, warlocks, and ghosts -- aka costumed kids. They shared a sample dataset with us. Each row represents a piece of candy that a costumed child was presented with during "trick" or "treat". We know if the candy was `chocolate` (or not chocolate) or `gummy` (or not gummy). Your goal is to predict if the costumed kid `ate` the piece of candy. 

If both chocolate and gummy equal one, you've got a chocolate gummy bear on your hands!?!?!
![Chocolate Gummy Bear](https://ed910ae2d60f0d25bcb8-80550f96b5feb12604f4f720bfefb46d.ssl.cf1.rackcdn.com/3fb630c04435b7b5-2leZuM7_-zoom.jpg)

In [1]:
import pandas as pd
candy = pd.read_csv('chocolate_gummy_bears.csv')

In [2]:
candy.head()

chocolate  gummy  ate
0          0      1    1
1          1      0    1
2          0      1    1
3          0      0    0
4          1      1    0

### Perceptron

To make predictions on the `candy` dataframe. Build and train a Perceptron using numpy. Your target column is `ate` and your features: `chocolate` and `gummy`. Do not do any feature engineering. :P

Once you've trained your model, report your accuracy. You will not be able to achieve more than ~50% with the simple perceptron. Explain why you could not achieve a higher accuracy with the *simple perceptron* architecture, because it's possible to achieve ~95% accuracy on this dataset. Provide your answer in markdown (and *optional* data anlysis code) after your perceptron implementation. 

In [3]:
# Start your candy perceptron here

X = candy[['chocolate', 'gummy']].values
y = candy['ate'].values

In [4]:
# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)


from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Perceptron by hand: 

model = Sequential()
model.add(Dense(1,input_dim=2, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X,y, epochs=5)

W0124 09:33:25.575685 140735661650816 deprecation.py:506] From /Users/michelle/opt/anaconda3/envs/U4-S2-NN/lib/python3.7/site-packages/tensorflow/python/ops/init_ops.py:1251: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
W0124 09:33:25.627903 140735661650816 deprecation.py:323] From /Users/michelle/opt/anaconda3/envs/U4-S2-NN/lib/python3.7/site-packages/tensorflow/python/ops/nn_impl.py:180: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


Epoch 1/5
10000/10000 [==============================] - 0s 40us/sample - loss: 0.6968 - acc: 0.5007
Epoch 2/5
10000/10000 [==============================] - 0s 30us/sample - loss: 0.6949 - acc: 0.5007
Epoch 3/5
10000/10000 [==============================] - 0s 41us/sample - loss: 0.6941 - acc: 0.5007
Epoch 4/5
10000/10000 [==============================] - 0s 33us/sample - loss: 0.6938 - acc: 0.5100
Epoch 5/5
10000/10000 [==============================] - 0s 36us/sample - loss: 0.6935 - acc: 0.5007


In [5]:
# evaluate the model
scores = model.evaluate(X, y)
print(f"{model.metrics_names[1]}: {scores[1]*100}")

10000/10000 [==============================] - 0s 32us/sample - loss: 0.6934 - acc: 0.5007
acc: 50.06999969482422


# Yep, this is only a mere 50% accuracy.

### Multilayer Perceptron <a id="Q3"></a>

Using the sample candy dataset, implement a Neural Network Multilayer Perceptron class that uses backpropagation to update the network's weights. Your Multilayer Perceptron should be implemented in Numpy. 
Your network must have one hidden layer.

Once you've trained your model, report your accuracy. Explain why your MLP's performance is considerably better than your simple perceptron's on the candy dataset. 

In [6]:
candy.head()

chocolate  gummy  ate
0          0      1    1
1          1      0    1
2          0      1    1
3          0      0    0
4          1      1    0

In [7]:
print(X.shape)
print(y.shape)

(10000, 2)
(10000,)


In [8]:
class NeuralNetwork:
    def __init__(self):
        # Set up Architecture of Neural Network
        self.inputs = 2
        self.hiddenNodes = 2
        self.outputNodes = 1

        # Initial Weights
        # 2x2 Matrix Array for the First Layer
        self.weights1 = np.random.rand(self.inputs, self.hiddenNodes)
       
        # 1 Matrix Array for Hidden to Output
        self.weights2 = np.random.rand(self.hiddenNodes, self.outputNodes)
        
    def sigmoid(self, s):
        return 1 / (1+np.exp(-s))
    
    def sigmoidPrime(self, s):
        sx = self.sigmoid(s)
        return sx * (1 - sx)
    
    def feed_forward(self, X_train):
        """
        Calculate the NN inference using feed forward.
        aka "predict"
        """
        
        # Weighted sum of inputs => hidden layer
        self.hidden_sum = np.dot(X_train, self.weights1)
        
        # Activations of weighted sum
        self.activated_hidden = self.sigmoid(self.hidden_sum)
        
        # Weight sum between hidden and output
        self.output_sum = np.dot(self.activated_hidden, self.weights2)
        
        # Final activation of output
        self.activated_output = self.sigmoid(self.output_sum)
        
        return self.activated_output
        
    def backward(self, X_train,y_train,o):#input matrix, output matrix and error
        """
        Backward propagate through the network
        """
        
        # Error in Output
        self.o_error = y_train - o
        
        # Apply Derivative of Sigmoid to error
        # How far off are we in relation to the Sigmoid f(x) of the output
        # ^- aka hidden => output
        self.o_delta = self.o_error * self.sigmoidPrime(o)
        
        # z2 error
        self.z2_error = self.o_delta.dot(self.weights2.T)
        
        # How much of that "far off" can explained by the input => hidden
        self.z2_delta = self.z2_error * self.sigmoidPrime(self.activated_hidden)
        
        # Adjustment to first set of weights (input => hidden)
        self.weights1 += X_train.T.dot(self.z2_delta)
        # Adjustment to second set of weights (hidden => output)
        self.weights2 += self.activated_hidden.T.dot(self.o_delta)
        

    def train(self, X_train, y_train):
        o = self.feed_forward(X_train)
        self.backward(X_train,y_train,o)

In [10]:
# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

#train the network 
import numpy as np

nn = NeuralNetwork()

# Number of Epochs / Iterations
for i in range(10000):
    if (i+1 in [1,2,3,4,5]) or ((i+1) % 1000 ==0):
        print('+' + '---' * 3 + f'EPOCH {i+1}' + '---'*3 + '+')
        print('Input: \n', X)
        print('Actual Output: \n', y)
        print('Predicted Output: \n', str(nn.feed_forward(X)))
        print("Loss: \n", str(np.mean(np.square(y - nn.feed_forward(X)))))
    nn.train(X,y)

+---------EPOCH 1---------+
Input: 
 [[0 1]
 [1 0]
 [0 1]
 ...
 [0 1]
 [0 1]
 [1 0]]
Actual Output: 
 [1 1 1 ... 1 1 1]
Predicted Output: 
 [[0.60293656]
 [0.60858398]
 [0.60293656]
 ...
 [0.60293656]
 [0.60293656]
 [0.60858398]]
Loss: 
 0.2611807657382774


ValueError: shapes (10000,10000) and (1,2) not aligned: 10000 (dim 1) != 1 (dim 0)

In [13]:
Acc = 1-.26118
Acc

0.73882

## Above we can see that the score increased to 73.9% Accuracy

In [11]:
model_improved = Sequential(name="Make it better")

model_improved.add(Dense(8, input_dim=2, activation='relu', name="Dense1"))
model_improved.add(Dense(8, activation='relu'))
model_improved.add(Dense(8, activation='relu'))
model_improved.add(Dense(1, activation='sigmoid'))

model_improved.compile(loss='binary_crossentropy', optimizer='adam',
              metrics=['accuracy'])

# Let's fit this
model_improved.fit(X,y, epochs=250, validation_split=.10)

Train on 9000 samples, validate on 1000 samples
Epoch 1/250
9000/9000 [==============================] - 1s 68us/sample - loss: 0.6079 - acc: 0.7212 - val_loss: 0.4752 - val_acc: 0.9400
Epoch 2/250
9000/9000 [==============================] - 0s 50us/sample - loss: 0.3542 - acc: 0.9464 - val_loss: 0.3236 - val_acc: 0.9400
Epoch 3/250
9000/9000 [==============================] - 0s 49us/sample - loss: 0.2883 - acc: 0.9464 - val_loss: 0.2990 - val_acc: 0.9400
Epoch 4/250
9000/9000 [==============================] - 0s 47us/sample - loss: 0.2692 - acc: 0.9464 - val_loss: 0.2854 - val_acc: 0.9400
Epoch 5/250
9000/9000 [==============================] - 1s 65us/sample - loss: 0.2565 - acc: 0.9464 - val_loss: 0.2710 - val_acc: 0.9400
Epoch 6/250
9000/9000 [==============================] - 1s 63us/sample - loss: 0.2465 - acc: 0.9464 - val_loss: 0.2609 - val_acc: 0.9400
Epoch 7/250
9000/9000 [==============================] - 0s 47us/sample - loss: 0.2384 - acc: 0.9464 - val_loss: 0.2546 - va

In [12]:
model_improved.evaluate(X,y)
#print(f"{model_improved.metrics_names[1]}: {scores[1]*100}")

10000/10000 [==============================] - 0s 22us/sample - loss: 0.2107 - acc: 0.9458


[0.21070679187774657, 0.9458]

## And now after running 250 ephochs and using a multilayer we have the Acc at 94.6%

P.S. Don't try candy gummy bears. They're disgusting. 

## 3. Keras MMP <a id="Q3"></a>

Implement a Multilayer Perceptron architecture of your choosing using the Keras library. Train your model and report its baseline accuracy. Then hyperparameter tune at least two parameters and report your model's accuracy.
Use the Heart Disease Dataset (binary classification)
Use an appropriate loss function for a binary classification task
Use an appropriate activation function on the final layer of your network.
Train your model using verbose output for ease of grading.
Use GridSearchCV or RandomSearchCV to hyperparameter tune your model. (for at least two hyperparameters)
When hyperparameter tuning, show you work by adding code cells for each new experiment.
Report the accuracy for each combination of hyperparameters as you test them so that we can easily see which resulted in the highest accuracy.
You must hyperparameter tune at least 3 parameters in order to get a 3 on this section.

In [17]:
import pandas as pd
from sklearn.preprocessing import StandardScaler

df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/heart.csv')
df = df.sample(frac=1)
print(df.shape)
df.head()

(303, 14)


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
125   34    0   1       118   210    0        1      192      0      0.7   
57    45    1   0       115   260    0        0      185      0      0.0   
282   59    1   2       126   218    1        1      134      0      2.2   
66    51    1   2       100   222    0        1      143      1      1.2   
114   55    1   1       130   262    0        1      155      0      0.0   

     slope  ca  thal  target  
125      2   0     2       1  
57       2   0     2       1  
282      1   1     1       0  
66       1   0     2       1  
114      2   0     2       1

In [18]:
X = df[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']]
y = df['target']

In [19]:
#train test split
import numpy as np
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.20, random_state=42)

In [25]:
print(X_train.shape)
print(y_train.shape)

(242, 13)
(242,)


In [ ]:
# Correct Encoding on Y
# What softmax expects = [0,0,0,0,0,1,0,0,0,0]

#y_train = keras.utils.to_categorical(y_train, num_classes)
#y_test = keras.utils.to_categorical(y_test, num_classes)

In [22]:
y_train[2]

1

In [23]:
# Hyper Parameters
#batch_size = 64
#num_classes = 10
#epochs = 20

In [24]:
from tensorflow import keras 
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

import numpy
import pandas as pd
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

# fix random seed for reproducibility
seed = 7
numpy.random.seed(seed)


In [26]:
# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

# Function to create model, required for KerasClassifier
def create_model():
    # create model
    model = Sequential()
    model.add(Dense(12, input_dim=13, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))#for binary classification
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)


# define the grid search parameters, leave epochs alone here
param_grid = {'batch_size': [10, 20, 40, 60, 80, 100],
              'epochs': [20]}

# Create Grid Search
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1)
grid_result = grid.fit(X, y) #don't need the split I did previously

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}") 

/Users/michelle/opt/anaconda3/envs/U4-S2-NN/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


Best: 0.6336633761723837 using {'batch_size': 10, 'epochs': 20}
Means: 0.6336633761723837, Stdev: 0.09181799468156064 with: {'batch_size': 10, 'epochs': 20}
Means: 0.6171617110570272, Stdev: 0.030605992514000077 with: {'batch_size': 20, 'epochs': 20}
Means: 0.5643564462661743, Stdev: 0.1220676159859578 with: {'batch_size': 40, 'epochs': 20}
Means: 0.5511551300684611, Stdev: 0.14581335187294317 with: {'batch_size': 60, 'epochs': 20}
Means: 0.4884488582611084, Stdev: 0.08644093203320917 with: {'batch_size': 80, 'epochs': 20}
Means: 0.5049504935741425, Stdev: 0.11317775187593769 with: {'batch_size': 100, 'epochs': 20}


In [27]:
# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

In [30]:
# Function to create a better model than above

from sklearn.model_selection import RandomizedSearchCV
def create_model():
    # create model
    model = Sequential()
    model.add(Dense(12, input_dim=13, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)


# define the grid search parameters, leave epochs alone here
param_distributions = {'batch_size': [10, 20, 40, 60, 80, 100],
              'epochs': [20]}

# Create Grid Search
grid = RandomizedSearchCV(estimator=model, param_distributions=param_distributions, n_jobs=1, verbose=1)
grid_result = grid.fit(X, y) #don't need the split I did previously

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}") 

Best: 0.6831683119138082 using {'epochs': 20, 'batch_size': 80}
Means: 0.6567656795183817, Stdev: 0.06969872060865748 with: {'epochs': 20, 'batch_size': 10}
Means: 0.646864672501882, Stdev: 0.02034459460670113 with: {'epochs': 20, 'batch_size': 20}
Means: 0.6501650015513102, Stdev: 0.032671590713967885 with: {'epochs': 20, 'batch_size': 40}
Means: 0.6501650214195251, Stdev: 0.05197366159998095 with: {'epochs': 20, 'batch_size': 60}
Means: 0.6831683119138082, Stdev: 0.02800424870987148 with: {'epochs': 20, 'batch_size': 80}
Means: 0.5907590985298157, Stdev: 0.07975608321765162 with: {'epochs': 20, 'batch_size': 100}


In [33]:
from sklearn.model_selection import RandomizedSearchCV
def create_model():
    # create model
    model = Sequential()
    model.add(Dense(12, input_dim=13, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)


# define the grid search parameters, try playing with Ephochs this time
param_distributions = {'batch_size': [80],
              'epochs': [100]}

# Create Grid Search
grid = RandomizedSearchCV(estimator=model, param_distributions=param_distributions, n_jobs=1, verbose=1)
grid_result = grid.fit(X, y) #don't need the split I did previously

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}") 

/Users/michelle/opt/anaconda3/envs/U4-S2-NN/lib/python3.7/site-packages/sklearn/model_selection/_search.py:266: UserWarning: The total space of parameters 1 is smaller than n_iter=10. Running 1 iterations. For exhaustive searches, use GridSearchCV.
  % (grid_size, self.n_iter, grid_size), UserWarning)
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 3 folds for each of 1 candidates, totalling 3 fits


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:   15.8s finished


Best: 0.7359736164410909 using {'epochs': 100, 'batch_size': 80}
Means: 0.7359736164410909, Stdev: 0.025986816922028384 with: {'epochs': 100, 'batch_size': 80}


## Increased the acc to 73.6% by tuning the layers, batch size, RandomizedSearchCV and then lastly the epochs.

In [34]:
# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

def create_model():
    # create model
    model = Sequential()
    model.add(Dense(12, input_dim=13, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(1, activation='sigmoid')) #for binary classification
    
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)


# define the grid search parameters, leave epochs alone here
param_grid = {'batch_size': [80, 100],
              'epochs': [100]}

# Create Grid Search
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1, verbose=1)
grid_result = grid.fit(X, y) 

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}") 

Fitting 3 folds for each of 2 candidates, totalling 6 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:   38.8s finished


Best: 0.6732673247655233 using {'batch_size': 80, 'epochs': 100}
Means: 0.6732673247655233, Stdev: 0.03523789943063056 with: {'batch_size': 80, 'epochs': 100}
Means: 0.646864672501882, Stdev: 0.030605992514000077 with: {'batch_size': 100, 'epochs': 100}


## Based on the above result with a lower acc of 67%, a RandomizedSearchCV works best